README.md
# nfe-calculos

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

**Note:** Replace ```Mário Sérgio Silvera``` ```dermarios``` ```https://about.me/dermarios``` ```dermarios@gmail.com``` ```gbbs``` ```nfe-calculos``` ```Cálculos de NFe``` with their correct values in [README.md](README.md), [CHANGELOG.md](CHANGELOG.md), [CONTRIBUTING.md](CONTRIBUTING.md), [LICENSE.md](LICENSE.md) and [composer.json](composer.json) files, then delete this line. You can run `$ php prefill.php` in the command line to make all replacements at once. Delete the file prefill.php as well.

This is where your description should go. Try and limit it to a paragraph or two, and maybe throw in a mention of what
PSRs you support to avoid any confusion with users and contributors.

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practices by being named the following.

```
bin/        
build/
docs/
config/
src/
tests/
vendor/
```


## Install

Via Composer

``` bash
$ composer require gbbs/nfe-calculos
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email dermarios@gmail.com instead of using the issue tracker.

## Credits

- [Mário Sérgio Silvera][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/gbbs/nfe-calculos.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/gbbs/nfe-calculos/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/gbbs/nfe-calculos.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/gbbs/nfe-calculos.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/gbbs/nfe-calculos.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/gbbs/nfe-calculos
[link-travis]: https://travis-ci.org/gbbs/nfe-calculos
[link-scrutinizer]: https://scrutinizer-ci.com/g/gbbs/nfe-calculos/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/gbbs/nfe-calculos
[link-downloads]: https://packagist.org/packages/gbbs/nfe-calculos
[link-author]: https://github.com/dermarios
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `nfe-calculos` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## NEXT - YYYY-MM-DD

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


tests/IPITest.php
<?php

declare(strict_types=1);

namespace Gbbs\NfeCalculos;

use PHPUnit\Framework\TestCase;

class IPITest extends TestCase
{
    /**
     * @expectedException \TypeError
     */
    public function testFailingNullArgument()
    {
        IPI::calcular(null);
    }

    /**
     * Test that object is Instance of IPI
     */
    public function testIsInstanceIPI()
    {
        $ipi = $this->instantiateIPI();
        $this->assertInstanceOf(IPI::class, $ipi);
    }

    /**
     * Test properties types
     */
    public function testPropertiesTypes()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCNPJProd('');
        $ipi->setCSelo('');
        $ipi->setQSelo(1.0);
        $ipi->setCEnq('');
        $ipi->setCST('');
        $ipi->setVBC(1.0);
        $ipi->setPIPI(1.0);
        $ipi->setVIPI(1.0);
        $ipi->setQUnid(1.0);
        $ipi->setVUnid(1.0);

        $this->assertIsString($ipi->getCNPJProd());
        $this->assertIsString($ipi->getCSelo());
        $this->assertIsFloat($ipi->getQSelo());
        $this->assertIsString($ipi->getCEnq());
        $this->assertIsString($ipi->getCST());
        $this->assertIsFloat($ipi->getVBC());
        $this->assertIsFloat($ipi->getPIPI());
        $this->assertIsFloat($ipi->getVIPI());
        $this->assertIsFloat($ipi->getQUnid());
        $this->assertIsFloat($ipi->getVUnid());
    }

    /**
     * Test invalid CST
     * @expectedException \Exception
     */
    public function testInvalidCST()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('00000');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 00
     * @expectedException \Exception
     */
    public function testCST00()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('00');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 01
     * @expectedException \Exception
     */
    public function testCST01()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('01');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 02
     * @expectedException \Exception
     */
    public function testCST02()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('02');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 03
     * @expectedException \Exception
     */
    public function testCST03()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('03');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 04
     * @expectedException \Exception
     */
    public function testCST04()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('04');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 05
     * @expectedException \Exception
     */
    public function testCST05()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('05');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 49
     * @expectedException \Exception
     */
    public function testCST49()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('49');

        IPI::calcular($ipi);
    }

    public function testCST50()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('50');
        $ipi->setVBC(1000.0);
        $ipi->setPIPI(12.0);

        IPI::calcular($ipi);

        $this->assertSame('50', $ipi->getCST());
        $this->assertSame(1000.0, $ipi->getVBC());
        $this->assertSame(12.0, $ipi->getPIPI());
    }

    public function testCST51()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('51');

        IPI::calcular($ipi);

        $this->assertSame('51', $ipi->getCST());
    }

    /**
     * Test CST 52
     * @expectedException \Exception
     */
    public function testCST52()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('52');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 53
     * @expectedException \Exception
     */
    public function testCST53()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('53');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 54
     * @expectedException \Exception
     */
    public function testCST54()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('54');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 55
     * @expectedException \Exception
     */
    public function testCST55()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('55');

        IPI::calcular($ipi);
    }

    /**
     * Test CST 99
     * @expectedException \Exception
     */
    public function testCST99()
    {
        $ipi = $this->instantiateIPI();
        $ipi->setCST('99');

        IPI::calcular($ipi);
    }

    /**
     * Instantiate and return an IPI object
     */
    private function instantiateIPI()
    {
        $ipi = new IPI();
        return $ipi;
    }
}


tests/PISTest.php
<?php

declare(strict_types=1);

namespace Gbbs\NfeCalculos;

use PHPUnit\Framework\TestCase;

class PISTest extends TestCase
{
    /**
     * @expectedException \TypeError
     */
    public function testFailingNullArgument()
    {
        PIS::calcular(null);
    }

    /**
     * Test that object is Instance of PIS
     */
    public function testIsInstancePIS()
    {
        $pis = $this->instantiatePIS();
        $this->assertInstanceOf(PIS::class, $pis);
    }

    /**
     * Test invalid CST
     * @expectedException \Exception
     */
    public function testInvalidCST()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('00000');

        PIS::calcular($pis);
    }

    /**
     * Test CST 01
     */
    public function testCST01()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('01');

        PIS::calcular($pis);

        $this->assertSame('01', $pis->getCST());
    }

    /**
     * Test CST 02
     * @expectedException \Exception
     */
    public function testCST02()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('02');

        PIS::calcular($pis);
    }

    /**
     * Test CST 03
     * @expectedException \Exception
     */
    public function testCST03()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('03');

        PIS::calcular($pis);
    }

    /**
     * Test CST 04
     * @expectedException \Exception
     */
    public function testCST04()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('04');

        PIS::calcular($pis);
    }

    /**
     * Test CST 05
     * @expectedException \Exception
     */
    public function testCST05()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('05');

        PIS::calcular($pis);
    }

    /**
     * Test CST 06
     * @expectedException \Exception
     */
    public function testCST06()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('06');

        PIS::calcular($pis);
    }

    /**
     * Test CST 07
     * @expectedException \Exception
     */
    public function testCST07()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('07');

        PIS::calcular($pis);
    }

    /**
     * Test CST 08
     * @expectedException \Exception
     */
    public function testCST08()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('08');

        PIS::calcular($pis);
    }

    /**
     * Test CST 09
     * @expectedException \Exception
     */
    public function testCST09()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('09');

        PIS::calcular($pis);
    }

    /**
     * Test CST 49
     * @expectedException \Exception
     */
    public function testCST49()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('49');

        PIS::calcular($pis);
    }

    /**
     * Test CST 50
     * @expectedException \Exception
     */
    public function testCST50()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('50');

        PIS::calcular($pis);
    }

    /**
     * Test CST 51
     * @expectedException \Exception
     */
    public function testCST51()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('51');

        PIS::calcular($pis);
    }

    /**
     * Test CST 52
     * @expectedException \Exception
     */
    public function testCST52()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('52');

        PIS::calcular($pis);
    }

    /**
     * Test CST 53
     * @expectedException \Exception
     */
    public function testCST53()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('53');

        PIS::calcular($pis);
    }

    /**
     * Test CST 54
     * @expectedException \Exception
     */
    public function testCST54()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('54');

        PIS::calcular($pis);
    }

    /**
     * Test CST 55
     * @expectedException \Exception
     */
    public function testCST55()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('55');

        PIS::calcular($pis);
    }

    /**
     * Test CST 56
     * @expectedException \Exception
     */
    public function testCST56()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('56');

        PIS::calcular($pis);
    }

    /**
     * Test CST 60
     * @expectedException \Exception
     */
    public function testCST60()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('60');

        PIS::calcular($pis);
    }

    /**
     * Test CST 61
     * @expectedException \Exception
     */
    public function testCST61()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('61');

        PIS::calcular($pis);
    }

    /**
     * Test CST 62
     * @expectedException \Exception
     */
    public function testCST62()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('62');

        PIS::calcular($pis);
    }

    /**
     * Test CST 63
     * @expectedException \Exception
     */
    public function testCST63()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('63');

        PIS::calcular($pis);
    }

    /**
     * Test CST 64
     * @expectedException \Exception
     */
    public function testCST64()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('64');

        PIS::calcular($pis);
    }

    /**
     * Test CST 65
     * @expectedException \Exception
     */
    public function testCST65()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('65');

        PIS::calcular($pis);
    }

    /**
     * Test CST 66
     * @expectedException \Exception
     */
    public function testCST66()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('66');

        PIS::calcular($pis);
    }

    /**
     * Test CST 67
     * @expectedException \Exception
     */
    public function testCST67()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('67');

        PIS::calcular($pis);
    }

    /**
     * Test CST 70
     * @expectedException \Exception
     */
    public function testCST70()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('70');

        PIS::calcular($pis);
    }

    /**
     * Test CST 71
     * @expectedException \Exception
     */
    public function testCST71()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('71');

        PIS::calcular($pis);
    }

    /**
     * Test CST 72
     * @expectedException \Exception
     */
    public function testCST72()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('72');

        PIS::calcular($pis);
    }

    /**
     * Test CST 73
     * @expectedException \Exception
     */
    public function testCST73()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('73');

        PIS::calcular($pis);
    }

    /**
     * Test CST 74
     * @expectedException \Exception
     */
    public function testCST74()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('74');

        PIS::calcular($pis);
    }

    /**
     * Test CST 75
     * @expectedException \Exception
     */
    public function testCST75()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('75');

        PIS::calcular($pis);
    }

    /**
     * Test CST 98
     * @expectedException \Exception
     */
    public function testCST98()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('98');

        PIS::calcular($pis);
    }

    /**
     * Test CST 99
     * @expectedException \Exception
     */
    public function testCST99()
    {
        $pis = $this->instantiatePIS();
        $pis->setCST('99');

        PIS::calcular($pis);
    }

    /**
     * Instantiate and return an PIS object
     */
    private function instantiatePIS()
    {
        $pis = new PIS();
        return $pis;
    }
}


prefill.php
<?php
define('COL_DESCRIPTION', 0);
define('COL_HELP', 1);
define('COL_DEFAULT', 2);

$fields = [
    'author_name' =>            ['Your name',             '',                                                ''],
    'author_github_username' => ['Your Github username',  '<username> in https://github.com/username',       ''],
    'author_email' =>           ['Your email address',    '',                                                ''],
    'author_twitter' =>         ['Your twitter username', '',                                                '@{author_github_username}'],
    'author_website' =>         ['Your website',          '',                                                'https://github.com/{author_github_username}'],

    'package_vendor' =>         ['Package vendor',        '<vendor> in https://github.com/vendor/package',   '{author_github_username}'],
    'package_name' =>           ['Package name',          '<package> in https://github.com/vendor/package',  ''],
    'package_description' =>    ['Package very short description',   '',                                     ''],

    'psr4_namespace' =>         ['PSR-4 namespace',       'usually, Vendor\\Package',                        '{package_vendor}\\{package_name}'],
];

$values = [];

$replacements = [
    'gbbs\\\\nfe-calculos\\\\' => function () use(&$values) { return str_replace('\\', '\\\\', $values['psr4_namespace']) . '\\\\'; },
    'Mário Sérgio Silvera'                 => function () use(&$values) { return $values['author_name']; },
    'dermarios'             => function () use(&$values) { return $values['author_github_username']; },
    'https://about.me/dermarios'              => function () use(&$values) { return $values['author_website'] ?: ('https://github.com/' . $values['author_github_username']); },
    'dermarios@gmail.com'                => function () use(&$values) { return $values['author_email'] ?: ($values['author_github_username'] . '@example.com'); },
    'gbbs'                      => function () use(&$values) { return $values['package_vendor']; },
    'nfe-calculos'                => function () use(&$values) { return $values['package_name']; },
    'Cálculos de NFe'         => function () use(&$values) { return $values['package_description']; },
    'League\\Skeleton'             => function () use(&$values) { return $values['psr4_namespace']; },
];

function read_from_console ($prompt) {
    if ( function_exists('readline') ) {
        $line = trim(readline($prompt));
        if (!empty($line)) {
            readline_add_history($line);
        }
    } else {
        echo $prompt;
        $line = trim(fgets(STDIN));
    }
    return $line;
}

function interpolate($text, $values)
{
    if (!preg_match_all('/\{(\w+)\}/', $text, $m)) {
        return $text;
    }
    foreach ($m[0] as $k => $str) {
        $f = $m[1][$k];
        $text = str_replace($str, $values[$f], $text);
    }
    return $text;
}

$modify = 'n';
do {
    if ($modify == 'q') {
        exit;
    }

    $values = [];

    echo "----------------------------------------------------------------------\n";
    echo "Please, provide the following information:\n";
    echo "----------------------------------------------------------------------\n";
    foreach ($fields as $f => $field) {
        $default = isset($field[COL_DEFAULT]) ? interpolate($field[COL_DEFAULT], $values): '';
        $prompt = sprintf(
            '%s%s%s: ',
            $field[COL_DESCRIPTION],
            $field[COL_HELP] ? ' (' . $field[COL_HELP] . ')': '',
            $field[COL_DEFAULT] !== '' ? ' [' . $default . ']': ''
        );
        $values[$f] = read_from_console($prompt);
        if (empty($values[$f])) {
            $values[$f] = $default;
        }
    }
    echo "\n";

    echo "----------------------------------------------------------------------\n";
    echo "Please, check that everything is correct:\n";
    echo "----------------------------------------------------------------------\n";
    foreach ($fields as $f => $field) {
        echo $field[COL_DESCRIPTION] . ": $values[$f]\n";
    }
    echo "\n";
} while (($modify = strtolower(read_from_console('Modify files with these values? [y/N/q] '))) != 'y');
echo "\n";

$files = array_merge(
    glob(__DIR__ . '/*.md'),
    glob(__DIR__ . '/*.xml.dist'),
    glob(__DIR__ . '/composer.json'),
    glob(__DIR__ . '/src/*.php'),
    glob(__DIR__ . '/tests/*.php')
);
foreach ($files as $f) {
    $contents = file_get_contents($f);
    foreach ($replacements as $str => $func) {
        $contents = str_replace($str, $func(), $contents);
    }
    file_put_contents($f, $contents);
}

echo "Done.\n";
echo "Now you should remove the file '" . basename(__FILE__) . "'.\n";


src/IPI.php
<?php

declare(strict_types=1);

namespace Gbbs\NfeCalculos;

use Exception;

class IPI
{
    public $descCST;
    public $clEnq;
    public $CodNCM;
    public $REIDE;
    public $SUFRAMA;
    public $CSTNotaRef;

    private $CNPJProd;
    private $cSelo;
    private $qSelo;
    private $cEnq;
    private $CST;
    private $vBC;
    private $pIPI;
    private $vIPI;
    private $qUnid;
    private $vUnid;

    /**
     * @param $IPI
     * @return mixed
     * @throws Exception
     */
    public static function calcular(IPI $IPI): void
    {
        if ($IPI->getCST() === '00') {
            /* ENTRADA COM RECUPERAÇÃO DE CRÉDITO */
            throw new Exception('Not implemented');
//            self::calcAliquotaAdValoren($IPI);
        } elseif ($IPI->getCST() === '01') {
            /* 01 ENTRADA TRIBUTADA COM ALICOTA ZERO */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '02') {
            /* ENTRADA ISENTA */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '03') {
            /* ENTRADA NÃO TRIBUTADA */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '04') {
            /* ENTRADA IMUNE */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '05') {
            /* ENTRADA COM SUSPENSAO */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '49') {
            /* OUTRAS ENTRADAS */
            throw new Exception('Not implemented');
//            self::calcRetornoIsentoValorOutros($IPI);
        } elseif ($IPI->getCST() === '50') {
            /* SAÍDA TRIBUTADA */
            self::calcAliquotaAdValoren($IPI);
        } elseif ($IPI->getCST() === '51') {
            /* SAÍDA TRIBUTADA COM ALICOTA ZERO */
            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '52') {
            /* SAÍDA ISENTA */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '53') {
            /* SAÍDA NÃO-TRIBUTADA */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '54') {
            /* SAÍDA IMUNE */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '55') {
            /* SAÍDA COM SUSPENSAO */
            throw new Exception('Not implemented');
//            self::calcIsento($IPI);
        } elseif ($IPI->getCST() === '99') {
            /* OUTRAS SAÍDAS */
            throw new Exception('Not implemented');
//            self::calcAliquotaAdValoren($IPI);
        } else {
            throw new Exception('Erro ao calcular IPI' . print_r($IPI, true));
        }
    }

    /**
     * ISENTO
     * @param $IPI
     */
    private static function calcIsento(self $IPI): void
    {
        $IPI->setVIPI(0.0);
        $IPI->setVBC(0.0);
        $IPI->setPIPI(0.0);
        $IPI->setQUnid(null);
        $IPI->setVUnid(null);
    }

    /**
     * Calcula o Valor IPI Ad Valoren
     * @param $IPI
     */
    private static function calcAliquotaAdValoren(self $IPI): void
    {
        $IPI->setVIPI($IPI->getVBC() * ($IPI->getPIPI() / 100));
    }

//    private static function calcRetornoIsentoValorOutros($IPI)
//    {
//        if ($IPI->CSTNotaRef === '00' || $IPI->CSTNotaRef === '50' || $IPI->CSTNotaRef === '99') {
//            $IPI->vIPI = ($IPI->vBC * ($IPI->pIPI / 100));
//            $IPI->pIPI = 0;
//        } else {
//            self::calcIsento($IPI);
//        }
//    }

    /**
     * @return string
     */
    public function getCNPJProd(): string
    {
        return $this->CNPJProd;
    }

    /**
     * @param string $CNPJProd
     */
    public function setCNPJProd(string $CNPJProd): void
    {
        $this->CNPJProd = $CNPJProd;
    }

    /**
     * @return string
     */
    public function getCSelo(): string
    {
        return $this->cSelo;
    }

    /**
     * @param string $cSelo
     */
    public function setCSelo(string $cSelo): void
    {
        $this->cSelo = $cSelo;
    }

    /**
     * @return float
     */
    public function getQSelo(): float
    {
        return $this->qSelo;
    }

    /**
     * @param float $qSelo
     */
    public function setQSelo(float $qSelo): void
    {
        $this->qSelo = $qSelo;
    }

    /**
     * @return string
     */
    public function getCEnq(): string
    {
        return $this->cEnq;
    }

    /**
     * @param string $cEnq
     */
    public function setCEnq(string $cEnq): void
    {
        $this->cEnq = $cEnq;
    }

    /**
     * @return string
     */
    public function getCST(): string
    {
        return $this->CST;
    }

    /**
     * @param string $CST
     */
    public function setCST(string $CST): void
    {
        $this->CST = $CST;
    }

    /**
     * @return float
     */
    public function getVBC(): float
    {
        return $this->vBC;
    }

    /**
     * @param float $vBC
     */
    public function setVBC(float $vBC): void
    {
        $this->vBC = $vBC;
    }

    /**
     * @return float
     */
    public function getPIPI(): float
    {
        return $this->pIPI;
    }

    /**
     * @param float $pIPI
     */
    public function setPIPI(float $pIPI): void
    {
        $this->pIPI = $pIPI;
    }

    /**
     * @return float
     */
    public function getVIPI(): float
    {
        return $this->vIPI;
    }

    /**
     * @param float $vIPI
     */
    public function setVIPI(float $vIPI): void
    {
        $this->vIPI = $vIPI;
    }

    /**
     * @return float
     */
    public function getQUnid(): float
    {
        return $this->qUnid;
    }

    /**
     * @param float $qUnid
     */
    public function setQUnid(?float $qUnid): void
    {
        $this->qUnid = $qUnid;
    }

    /**
     * @return float
     */
    public function getVUnid(): float
    {
        return $this->vUnid;
    }

    /**
     * @param float $vUnid
     */
    public function setVUnid(?float $vUnid): void
    {
        $this->vUnid = $vUnid;
    }
}


src/COFINS.php
<?php

declare(strict_types=1);

namespace Gbbs\NfeCalculos;

use Exception;

class COFINS
{
    private $CST;
    public $vBC;
    public $pCOFINS;
    public $vCOFINS;
    public $pRedBC;
    public $vAliqProd;
    public $qBCProd;
    public $Desconto;
    public $REIDE;
    public $SUFRAMA;
    public $Zerar;

    /**
     * @param $COFINS
     * @return mixed
     * @throws Exception
     */
    public static function calcular(COFINS $COFINS)
    {
        /* Operação Tributável com Alíquota Básica */
        if ($COFINS->getCST() === '01') {
            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '02') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '03') {
            throw new Exception('Not implemented');
//            return self::calcAliqCST03($COFINS);
        } elseif ($COFINS->getCST() === '04') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '05') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '06') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '07') {
            throw new Exception('Not implemented');
//            return self::calcIsentoDesconto($COFINS);
        } elseif ($COFINS->getCST() === '08') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '09') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '49') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '50') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '51') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '52') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '53') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '54') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '55') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '56') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '60') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '61') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '62') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '63') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '64') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '65') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '66') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '67') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '70') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        } elseif ($COFINS->getCST() === '71') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '72') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '73') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '74') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '75') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '98') {
            throw new Exception('Not implemented');
//            return self::calcaCOFINS($COFINS);
        } elseif ($COFINS->getCST() === '99') {
            throw new Exception('Not implemented');
//            return self::calcIsento($COFINS);
        }
        throw new Exception('Erro ao calcular COFINS' . print_r($COFINS, true));
    }

//    private static function calcIsento($COFINS)
//    {
//        $COFINS->pCOFINS = 3;
//        if ($COFINS->REIDE === 1 || $COFINS->SUFRAMA === 1) {
//            $COFINS->Desconto = $COFINS->vBC * ($COFINS->pCOFINS / 100);
//        } else {
//            $COFINS->Desconto = 0;
//        }
//        $COFINS->vCOFINS = '0';
//        $COFINS->vBC = '0';
//        $COFINS->pCOFINS = '0';
//
//        return $COFINS;
//    }

    private static function calcaCOFINS($COFINS)
    {
        $COFINS->pCOFINS = 3;
        $COFINS->pRedBC = 0;
        $COFINS->vCOFINS = (($COFINS->vBC - ($COFINS->vBC * ($COFINS->pRedBC / 100))) * ($COFINS->pCOFINS / 100));
        if ($COFINS->REIDE === 1 || $COFINS->SUFRAMA === 1) {
            $COFINS->Desconto = $COFINS->vBC * ($COFINS->pCOFINS / 100);
        } else {
            $COFINS->Desconto = 0;
        }
        return $COFINS;
    }

//    private static function calcIsentoDesconto($COFINS)
//    {
//        $COFINS->pCOFINS = 3;
//        if ($COFINS->REIDE === 1 || $COFINS->SUFRAMA === 1) {
//            $COFINS->Desconto = $COFINS->vBC * ($COFINS->pCOFINS / 100);
//        } else {
//            $COFINS->Desconto = 0;
//        }
//
//        $COFINS->pCOFINS = '0';
//        $COFINS->vBC = '0';
//        $COFINS->pCOFINS = '0';
//        $COFINS->vCOFINS = '0';
//
//        return $COFINS;
//    }
//
//    private static function calcAliqCST03($COFINS)
//    {
//        $COFINS->vBC = 0;
//        $COFINS->pCOFINS = 0;
//        $COFINS->vCOFINS = round((($COFINS->vAliqProd) * ($COFINS->qBCProd)), 2);
//        $COFINS->Desconto = 0;
//
//        return $COFINS;
//    }

    /**
     * @return string
     */
    public function getCST(): string
    {
        return $this->CST;
    }

    /**
     * @param string $CST
     */
    public function setCST(string $CST): void
    {
        $this->CST = $CST;
    }
}
